[Security] Added tests to check referer URL with a login_path route name

wcluijt opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | maybe?
| Deprecations? | no
| Tests pass?   | no
| Fixed tickets | none
| License       | MIT
| Doc PR        | N/A

Related to https://github.com/symfony/symfony/pull/19026

wcluijt committed:
Added tests to check referer URL with a login_path route name

nicolas-grekas commented:
@wcluijt can you please explain why you submitted this PR? Can you check #23061 also, is it related, how for you?

wcluijt commented:
@nicolas-grekas In my application, I have some Behat tests which use Selenium and relate to form login, which were not passing after upgrading to Symfony 3.3.0.  In those tests, the referer value in the `$targetUrl` variable was a full URL.  Also in my application, I have the firewall `form_login.login_path` specified with a route name in my `app/config/security.yml` file, which will produce a URL in the `$this->httpUtils->generateUri(...)` call of the `DefaultAuthenticationSuccessHandler`.  After upgrading to Symfony 3.3.0, when I manually checked the login form in my browser, instead of redirecting me to a different page, my application redirected me back to the login form page (I appeared to be logged in as well, but just ended up on the wrong page).

I added these tests to document functionality I believe existed well before https://github.com/symfony/symfony/pull/19026 was merged (at least for the first test I added).  For that particular PR, I believe that the introduction of the `parse_url(...)` call satisfied the requirements of the underlying issue presented in https://github.com/symfony/symfony/issues/18862 to remove query parameters in the referer value, but may have inadvertently removed the URL protocol and domain as well (if there was one).  I suppose that part of it wasn't caught at the time because there were no tests checking for that.

Yes, I think https://github.com/symfony/symfony/pull/23061 relates directly, but rather than adding yet another call to `parse_url(...)`, maybe the first one around `$targetUrl` can be re-evaluated to consider these tests and the functionality refactored to only remove query parameters, without touching any other part of the value.  I suppose I'm thinking of the case (however unlikely) where the referer is a URL like `http://example.net/login` but the generated URL based on `login_path` with a route name is `http://example.com/login`, where the domain in this case makes a difference.

fabpot commented:
Fix in 19026 is wrong. So, instead, I've submitted #23580, which fixed the issue in 2.7 as a bug fix. When merged up to master, that would fix the issue in 3.3 as well. Closing in favor of #23580.
